docs/extensibility/extend-rest-api-add-data.md

```diff
@@ -0,0 +1,289 @@
+# Exposing your data in the Store API.
+
+## The problem
+You want to extend the Cart and Checkout blocks, but you want to use some custom data not available on Store API or the context.
+You don't want to create your own endpoints or Ajax actions. You want to piggyback on the existing StoreAPI calls.
+
+## Solution
+ExtendRestAPI offers the possibility to add contextual custom data to Store API endpoints, like `wc/store/cart` and `wc/store/cart/items` endpoints.
+That data is namespaced to your plugin and protected from other plugins malfunctioning.
+The data is available on all frontend filters and slotFills for you to consume.
+
+## Basic usage
+You can use ExtendRestAPI by registering a couple of functions, `schema_callback` and `data_callback` on a specific endpoint namespace. Those functions are going to be called at execution time and are going to be passed relevant data.
+
+```PHP
+
+use Automattic\WooCommerce\Blocks\Package;
+use Automattic\WooCommerce\Blocks\Domain\Services\ExtendRestApi;
+use Automattic\WooCommerce\Blocks\StoreApi\Schemas\CartSchema;
+
+add_action('woocommerce_blocks_loaded', function() {
+ // ExtendRestApi relies on a Singleton that is shared between the API and consumers.
```

Comment:
That's a great idea!

Reply:
cc @nerrad because my PHP and DI knowledge is rusty.

Is it possible for us to have `new ExtendRestApi` still fetch the same instance as `Package::container()->get( ExtendRestApi::class )`?
I think I saw patterns like this where `ExtendRestApi` constructor would just assign its return value to whatever is in `Package::container()->get( ExtendRestApi::class )`.

This would require us to change two things:
- Change ExtendRestApi so it's initiated with `init` function, calling `__construtor` would just return whatever `Package::container()->get( ExtendRestApi::class )`.
```php
public function __construct() {
	return Package::container()->get( self::class );
}

public function init( Package $package, Formatters $formatters ) {
	$this->package    = $package;
	$this->formatters = $formatters;
}
```
- This code block in `Domain/Bootstrap.php` would accommodate that change somehow
```php
$this->container->register(
	ExtendRestApi::class,
	function( Container $container ) {
		return new ExtendRestApi( $container->get( Package::class ), $container->get( Formatters::class ) );
	}
);
```